### ISSUE TITLE
Using ngCordova to set up the facebook authentication

### ISSUE THREAD

**Issue by casertap** [lynndylanhurley/ng-token-auth#141]
I would like to use ng-token-auth for the email authentication but I would like to use ngCordova for the facebook oauth authentication.
Everything work fine with email and cordova oauth but I do not know how to update ng-token-auth with the access_token so it knows the user is authenticated.

My code look like this:
```
 $cordovaOauth.facebook("9...4", ["email"]).then(function(result) {
            console.log("CONNECTED");
            console.log(result);           
        }, function(error) {
            console.log("ERROR");
            console.log(error);
        });
````
The output look like this after authentication:
````
CONNECTED
{"access_token":"CAANOTIKs5WABAKPF7DZC8M1n1U7QmsYd9IjmdAl7RHVhgwHveE83X0ZCRVsO6ZA4HUt3478XB...yHZAZBoB2gty03tL6b4zyiYxcCB6dOjb03R8mmfeIXR5yupx74alcpZBlAx5pWRslDOsjfM2MZD","expires_in":"5182962"}
`````

How can I set the ng-token-auth access_token? 
How to call the handleValidAuth method ? 
Do I need to perform an event broadcast using $rootScope.$broadcast('auth:login-success', ??? ); ?

**Comment by sebastianzillessen** [comment]
Hi, 

I have exactly the same issue, I have a user authetified by the Facebook Cordova Plugin (native auth flow) and received a Access-token, and now I want to use it to register at the server. How can this be done?

**Comment by posixpascal** [comment]
+1 same problem as well.

**Comment by heysailor** [comment]
ng-token-auth is set up for the server to obtain the FB access_token server-side directly from FB, using server redirects and a callback from Facebook back to the server.

ngCordova/facebookConnect plugins obtain the access_token on the client side, using either the native app or a browser window.

To use ng-token-auth with these cordova plugins, a new function has to be added to ng-token-auth.

Essentially, when you have the FB access_token from the success function of $cordovaFacebook, you hand it to ng-token-auth - which passes this client-side FB access_token to server. The server then itself then needs an API to handle verify this provided FB access_token. If it's a viable FB access_token, then the server will need to return a ng-token-auth acces_token to the client, to say all is well.

Hope that helps. I'm implementing the same myself, and will post a link when I'm done.

**Comment by heysailor** [comment]
Have a look at: https://github.com/lynndylanhurley/ng-token-auth#oauth2-authentication-flow

This shows the flow currently in ng-token-auth for server side OAuth.

Bear in mind we have two types of access_token flying around: the Facebook one, and the one we want generated via ng-token-auth to allow access to OUR client app. They are not interchangeable.

When dealing with ngCordova's $facebookConnect, the concepts of the first three boxes all occur client side. $facebookConnect provides the API of the second (red) box. The 'redirect to redirect URI' is the localhost callback that $facebookConnect uses, to receive the FB access_token from the native Facebook app.

We need to implement a client step (green box) right after this redirect step, which passes the FB access token to the server API. The server API steps will be different, to verify the passed FB access_token server-side. The outcome will be similar, a response of a access_token for access to our app (NOT the facebook token). It will be a regular http response, not a postmessage.

**Comment by heysailor** [comment]
Here's my fork: https://github.com/heysailor/ng-token-auth-localOAuth
And on bower: bower install --save ng-token-auth-localOAuth

**Comment by posixpascal** [comment]
I'd contribute as well, we'll also have to modify the rails-gem. I'm fiddling around with it after work.

**Comment by heysailor** [comment]
Whoops bower is: bower install --save ng-token-auth-local-oauth#0.0.25-oauth

**Comment by heysailor** [comment]
and I made it https://github.com/heysailor/ng-token-auth-local-oauth for clarity.

**Comment by lynndylanhurley** [comment]
@heysailor - is this something that can work alongside the existing non-cordova functionality? In other words, is this something that can be merged into this repo without any breaking changes?

**Comment by posixpascal** [comment]
I tested it, it works alongside current ng-token-auth, doesn't break anything so far.

It doesn't work with the devise_token_auth gem (yet), so you'd have to write your own oauth-validator. Maybe we can work together and add cordova support to the gem as well :+1:

**Comment by lynndylanhurley** [comment]
> Maybe we can work together and add cordova support to the gem as well

Yeah I'm down! I don't have much experience with this form of authentication - would you like to take the lead on this feature?

**Comment by posixpascal** [comment]
yep, sure.
I'm gonna fork your gem & play around with it.

**Comment by heysailor** [comment]
Yes it's working for me alongside the other code too. I'm using sails as a
backend so can't do the gem updating. But having this addition to
ng-token-auth will help all the Cordova users out there.

I added some documentation to the readme too.

**Comment by lynndylanhurley** [comment]
@posixpascal - cool thx!!! This is by far the most requested feature. If we get this working I think I'll be able to close half of the issues on devise_token_auth :smiley:

**Comment by heysailor** [comment]
Should be fairly straightforward....the hard work of obtaining the oauth
token is done already client side.

For example, I do a server side query to the facebook graph token debug
feature to verify that the access token is valid before creating or
authenticating that user.

**Comment by sebastianzillessen** [comment]
Any updates on the Server side (especially the `devise-token-auth` gem)?

**Comment by posixpascal** [comment]
I couldn't work on it during the last week. I'm unfortunately very busy at the moment, I try to make ionic work with devise-token-auth gem in my spare time.

**Comment by davidpatrick** [comment]
Good work guys.  A couple months ago I solved this problem without any big changes to this gem or devise-token-auth, basically just a small change on the redirect page.  https://github.com/lynndylanhurley/ng-token-auth/issues/90#issuecomment-83058273

The main difference in approach is that here we are allowing something other than our server to accept the token, and then injecting that token into our server.

You can see in the other thread I basically achieve what the ng-cordova-oauth plugin does but instead of grabbing the token from localhost I grab it from our devise-token-auth page and emit it into ng-token-auth.  We could just as easily shape our redirect page at devise-token-auth around the ng-cordova-oauth localhost, and then do a PR to ng-cordova-oauth that would make the redirect uri configurable.

**Comment by posixpascal** [comment]
Alright guys. I made it :). Big thanks to davidpatrick for his example on #90. I changed his code slightly to reflect the current version of ng_token_auth gem. It works using the cordovaInAppBrowser plugin:

`cordova plugin add https://github.com/apache/cordova-plugin-inappbrowser`

You have to use the github mirror because `cordova plugin add cordova-plugin-inappbrowser` fails.

I will post a complete example within the next 2 days & make a PR to this repo & the devise_token_auth gem. :).

:+1:  keep up the good work community.

**Comment by casertap** [comment]
Awesome @posixpascal ! Thanks a lot.
I will test your that as soon as the doc is published. :+1:

**Comment by krk324** [comment]
Great @posixpascal ! Thank you very much!
Looking forward for the complete example and repo :)

**Comment by kabus202** [comment]
is it published already? :+1:

**Comment by posixpascal** [comment]
sup, not yet unfortunately. I still have lots of work to do and don't find time to push changes to the ng-token-auth repo. I'll look into it as soon as I have a little spare time.

**Comment by krk324** [comment]
Any updates on the Server side (especially the devise-token-auth gem)? Trying to figure it out myself but no progress so far.

**Comment by posixpascal** [comment]
I got it working without ngCordova (just noticed that this issue is about ngCordova's Facebook Plugin).
If you can live with using the InAppBrowser approach I can show you how to do it.

First you have to make sure you get the cordova-inappbrowser plugin and included $ngCordova in your `angular.module("yourapp", [/*HERE*/]` modules array.

Then you have to overwrite `omniauth_response.html.erb` in your rails app.
I created the file for my own here: _$RAILS_ROOT/app/views/layouts/omniauth_response.html.erb`

Rails
====
I use monkey-patching to overwrite the gem-html file without touching the gem itself.
The content is mostly the same as the regular one:

```ruby
<!DOCTYPE html>
<html>
  <head>
    <script>
    	 var getCredentials = function(){
          return {
            <%= yield %>
          };
        };

      var usingExternalWindow = function() {
        var nav     = navigator.userAgent.toLowerCase(),
            ieLTE10 = (nav && nav.indexOf('msie') != -1),
            ie11    = !!navigator.userAgent.match(/Trident.*rv\:11\./);
        return !(ieLTE10 || ie11);
      }
      if (usingExternalWindow()) {
	       // browser flow.
        window.addEventListener("message", function(ev) {

          if (ev.data === "requestCredentials") {
            ev.source.postMessage(getCredentials(), "*");
            window.close();
          }
        });
      } else {

        window.location.href = "<%= @auth_origin_url %>";
      }
    </script>
  </head>
  <body>
    <pre>
      Redirecting...
    </pre>
  </body>
</html>
```

Only difference is that I extracted the <%= yield %> part in an external function. 
I used @davidpatrick example on #90.

ionic / cordova / angular
==================
Next thing you have to do is write a custom service for handling cordova auth:
My login service looks like this:

```javascript
angular.module('app.services.login', [])
.service('loginService', ['$rootScope', '$q', '$window', '$auth',
  function ($rootScope, $q, $window, $auth) {
    var _this = this;

    this.originUrl = $window.location.origin;
    var loginDomain = "http://frood.herokuapp.com";
    this.authUrl = loginDomain  + "/auth/";

    this.authenticate = function(provider) {
      var defer = $q.defer();
      console.log(_this);
      console.log(provider);
      var browserWindow = $window.open(_this.authUrl + provider + "?auth_origin_url=" + this.originUrl, '_blank', 'location=no');

      browserWindow.addEventListener('loadstop', function(event) {
        if (event.url.match( loginDomain )) {
          browserWindow.executeScript(
            { code:'getCredentials()' },

            function(response){
              var rsp = response[0];

              if (rsp.message == 'deliverCredentials') {
                var user = rsp;

                // Set the user in ng-token-auth
                $auth.initDfd();
                $auth.handleValidAuth(user, true);

                // emit a successful login
                $rootScope.$emit('auth:login-success', user)
                defer.resolve(user);
              } else {
                $rootScope.$emit('auth:login-error', rsp.error)
                defer.reject(rsp.error);
              };

              // Finished with InAppBrowser
              browserWindow.close();
            }
          );
        };
      });

      return defer.promise;
    };
  }
]);
```

Then in my AuthController I do this:
```javascript
$scope.authenticateWith = function(what){
      var auth = ("cordova" in window) ? loginService : $auth;
      auth.authenticate(what).then(function(){
        if (typeof $rootScope.modal !== "undefined"){
          $rootScope.modal.hide();
        }
      });
```

This got it working on mobile phones flawless, well not 100% flawless because it uses the web-version of oauth providers (eg. not the native app), so people aren't logged in when they never used the web version.

I'm making a pull request soon to add inAppBrowser support out of the box. 
I got a lot of work to do at the moment and I'm not able to contribute much :(

There is a way to use $ngCordova as well but I haven't looked into it much.

**Comment by tohyongcheng** [comment]
Anyone has tried ngCordova with devise-token-auth? I have been trying to work around it, by hacking with the access code and pushing them the callback endpoint. I enable this by changing the strategy file to accept a access_token parameter in the request. More information can be found here: https://github.com/mkdynamic/omniauth-facebook/pull/174

So what I do is:
- use ngCordova $cordovaFacebook to login, and get the access_token
- call /auth/facebook/callback?access_token=12354671298387123921
- try to receive it in omniauth_callbacks_controller.rb

However, I failed at the last step and am unable to get the proper auth_hash to be returned. I hope that these information can prove useful to someone trying to achieve the same thing and help in finding a solution!

**Comment by posixpascal** [comment]
I don't know exactly how devise works but maybe you have to open "/auth/facebook" first & then use ngCordova to get the access token and push it to the facebook callback method.

**Comment by booleanbetrayal** [comment]
@posixpascal / anyone-who-is-willing: I'd love to get Cordova OAuth support baked into this module and reduce the number of custom forks that people are requiring. Can you you provide a formal PR for me to take a look at? If so, it'll get top-priority. Same applies for devise_token_auth.

**Comment by booleanbetrayal** [comment]
@nbrustein and I are working on a refactoring of the omniauth controller logic in order to support `sameWindow` *(new default)*, `externalWindow`, and `inAppBrowser` flows. We should have a PR soon (hopefully).

**Comment by booleanbetrayal** [comment]
The PRs are in for ng-token-auth #188 and devise-token-auth (lynndylanhurley/devise_token_auth#323). Please try these branches out and let us know if there are any issues in implementation. Will merge soon if no major issues are found!

**Comment by booleanbetrayal** [comment]
Should be closed via #188. Let me know if you guys run into any issues!
